lib/did/did.go

```diff
@@ -0,0 +1,336 @@
+package did
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"github.com/btcsuite/btcutil/base58"
+
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/ed25519"
+	"golang.org/x/crypto/nacl/box"
+)
+
+var (
+	ErrInvalidSeed       = errors.New("invalid seed")
+	ErrInvalidDID        = errors.New("invalid DID")
+	ErrInvalidPubKeySize = errors.New("invalid public key size")
+	ErrDecryptionFailed  = errors.New("decryption failed")
+	ErrInvalidSignature  = errors.New("invalid signature")
+	ErrInvalidEncoding   = errors.New("invalid encoding")
+	ErrSerialization     = errors.New("serialization error")
+	ErrDeserialization   = errors.New("deserialization error")
+)
+
+// decentralized identifier (string wrapper type)
+//
+// takes the form of "did:key:<base58-encoded ed25519 public key>"
+type DID string
+
+// returns the inner string of the DID
+func (d DID) String() string {
+	return string(d)
+}
+
+// a DID provider that uses the ed25519 signature algorithm
+type Ed25519Provider struct {
+	pubKey  ed25519.PublicKey
+	privKey ed25519.PrivateKey // no public "getter" for priv key
+	did     DID
+}
+
+// getter for the DID
+func (p *Ed25519Provider) DID() DID {
+	return p.did
+}
+
+// getter for the public key
+func (p *Ed25519Provider) PublicKey() ed25519.PublicKey {
+	return p.pubKey
+}
+
+// creates a new DID provider using the seed as the private key
+//
+// takes a 32-byte high-entropy seed
+func NewEd25519Provider(seed []byte) (*Ed25519Provider, error) {
```

Comment:
```suggestion
func NewEd25519Provider(seed [ed25519.SeedSize]byte) (*Ed25519Provider, error) {
```

Better just to use type checking rather than a slice with a runtime check. Note: This means you need to update all the tests with the appropriate type.